src/icp_canister_backend/src/lib.rs

```diff
@@ -62,54 +69,103 @@ pub async fn deposit(user: Principal, timelock: u64) -> Result<(), types::Deposi
         "icrc1_balance_of",
-        (account,)
+        (from_account.clone(),)
     ).await;
 
     let balance = match balance_result {
         Ok((balance,)) => balance,
         Err(_) => return Err(types::DepositError::LedgerCallFailed),
     };
 
-    let balance_u64: u64 = balance.0.try_into().map_err(|_| types::DepositError::InternalError)?;
+    let balance_u64: u64 = balance.0.clone().try_into().map_err(|_| types::DepositError::InternalError)?;
 
     if balance_u64 == 0 {
         return Err(types::DepositError::InsufficientBalance);
     }
+
+    // Calculate transfer amount (balance minus fee)
+    // First, let's get the transfer fee from the ledger
+    let fee_result: Result<(Nat,), _> = call(
+        ledger_principal,
+        "icrc1_fee",
+        ()
+    ).await;
     
-    let deposit = types::Deposit {
-        deposit_time: ic_cdk::api::time(),
-        user,
-        subaccount: subaccount.to_vec(),
-        amount: balance_u64,
-        timelock,
+    let transfer_fee = match fee_result {
+        Ok((fee,)) => fee,
+        Err(_) => Nat::from(10_000u64), // Default fee if call fails
     };
     
-    let deposit_json = serde_json::to_string(&deposit).expect("Failed to serialize");
-    DEPOSITS.with(|deposits| {
-        deposits.borrow_mut().insert(deposit_key, deposit_json);
-    });
+    // Calculate amount to transfer (balance minus fee)
+    let transfer_amount = if balance.0 > transfer_fee.0 {
+        Nat::from(&balance.0 - &transfer_fee.0)
+    } else {
+        return Err(types::DepositError::InsufficientBalance); // Not enough for fee
+    };
+
+    // Transfer tokens from subaccount to canister's main account
+    let transfer_args = (
+        types::TransferArg {
+            from_subaccount: Some(subaccount.to_vec()),
+            to: types::Account {
+                owner: ic_cdk::api::id(),
+                subaccount: None, // Main account
+            },
+            amount: transfer_amount.clone(),
+            fee: Some(transfer_fee),
+            memo: Some(b"deposit_transfer".to_vec()),
+            created_at_time: None,
+        },
+    );
 
-    Ok(())
+    let transfer_result: Result<(Result<Nat, types::TransferError>,), _> = call(
+        ledger_principal,
+        "icrc1_transfer",
+        transfer_args
+    ).await;
+
+    match transfer_result {
+        Ok((Ok(_block_index),)) => {
+            // Transfer successful, create deposit entry
+            let current_time = ic_cdk::api::time();
+            let unlock_time = current_time + (timelock * 1_000_000_000); // Convert seconds to nanoseconds
+            
+            let deposit = types::Deposit {
+                unlocktime: unlock_time,
+                principal: user,
+                amount: transfer_amount.0.try_into().map_err(|_| types::DepositError::InternalError)?,
+            };
+
+            // Get current counter and increment it
+            let deposit_id = DEPOSIT_COUNTER.with(|counter| {
```

Comment:
Rearrange steps in more natural flow:
First incriment deposit_counter and get deposit_id
Then create and store Deposit struct.